ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,612 @@
+        case OSSL_QUIC_FRAME_TYPE_HANDSHAKE_DONE: {
+            if (!ossl_quic_wire_decode_frame_handshake_done(pkt))
+                return 0;
+
+            switch (mode) {
+            case collect_ack_data:
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+                break;
+            case process_frames: {
+                /* ADD CODE to tell the handshake manager that we're done */
+                break;
+            }
+            }
+            break;
+        }
+        }
+    }
+
+    return 1;
+}
+
+static int ossl_quic_handle_frames(QUIC_CONNECTION *connection,
+                                   OSSL_QRX_PKT *qpacket)
+{
+    PACKET pkt;
+    OSSL_ACKM_RX_PKT ackm_data;
+    int ok = 0;                  /* Assume the worst */
+
+    /* Phase 0: special case packet types that have no payload */
+    if (qpacket->hdr->type == QUIC_PKT_TYPE_RETRY) {
+        /* ADD CODE to handle a retry */
+        goto success;
+    }
+
+    /* Phase 1: go through all frames and collect ACK manager data */
+    memset(&ackm_data, 0, sizeof(ackm_data));
+    /* ASSUMPTION: All packets that aren't special case have a packet number */
+    ackm_data.pkt_num = qpacket->pn;
+    ackm_data.time = qpacket->time;
+    switch (qpacket->hdr->type) {
+    case QUIC_PKT_TYPE_INITIAL:
+    case QUIC_PKT_TYPE_VERSION_NEG:
+        ackm_data.pkt_space = QUIC_PN_SPACE_INITIAL;
+        break;
+    case QUIC_PKT_TYPE_HANDSHAKE:
+        ackm_data.pkt_space = QUIC_PN_SPACE_HANDSHAKE;
+        break;
+    case QUIC_PKT_TYPE_0RTT:
+    case QUIC_PKT_TYPE_1RTT:
+        ackm_data.pkt_space = QUIC_PN_SPACE_APP;
+        break;
+    }
+
+    ackm_data.is_ack_eliciting = 0;
+    if (!PACKET_buf_init(&pkt, qpacket->hdr->data, qpacket->hdr->len)
+        || !depack_process_frames(NULL, &pkt, ackm_data.pkt_space,
+                                  qpacket->time, collect_ack_data, &ackm_data))
+        goto end;
+
+    /* ASSUMPTION: The connection object has an ACK manager attached */
+    if (!ossl_ackm_on_rx_packet(GET_CONN_ACKM(connection), &ackm_data))
+        goto end;
+
+    /* Phase 2: go through all frames and process their payloads */
+    if (!PACKET_buf_init(&pkt, qpacket->hdr->data, qpacket->hdr->len)
+        || !depack_process_frames(connection, &pkt, ackm_data.pkt_space,
+                                  qpacket->time, process_frames, NULL))
+        goto end;
+
+ success:
+    ok = 1;
+ end:
+    /* Placeholder for mandatory cleanups */
+    return ok;
+}
+
+int ossl_quic_depacketize(QUIC_CONNECTION *connection)
+{
+    OSSL_QRX_PKT qpacket;
+    int rc;
+
+    /* Try to read a packet from the read record layer */
+    memset(&qpacket, 0, sizeof(qpacket));
+    if ((rc = ossl_qrx_read_pkt(GET_CONN_QRX(connection), &qpacket)) <= 0)
+        return 0;
+
+    rc = ossl_quic_handle_frames(connection, &qpacket);
+    ossl_qrx_release_pkt(GET_CONN_QRX(connection), qpacket.handle);
```

Comment:
I don't think it is so simple. See my comment above about `ssl_queue_data` and ref counting the `OSSL_QRX_PKT`. In my suggested solution only when the ref count on the `OSSL_QRX_PKT` drops to 0 would we call `ossl_qrx_release_pkt`.

Reply:
I assume that `ossl_qrx_release_pkt()` can handle the refcount decrement.
(it could also be modified to take a pointer to `OSSL_QRX_PKT` rather than the mysterious opaque pointer that it holds...  @hlandau?)

Reply:
I've addressed the parameter question: https://github.com/openssl/openssl/pull/18949/files#r947918011)

Reply:
This has been fixed with a wrapper